README.md
# ex_mustang

> A simple, clueless bot

![Mustang](images/mustang.jpg)

ExMustang is a bot for Slack written in Elixir. While ExMustang is written for Slack, the responder modules should work fine with other Hedwig adapters as well. If you like Mustang, please use [Mustang](images/mustang.jpg) as bot image if possible :)

_Warning: This is a work in progress._

![Example](images/example.png)

![Example](images/example2.png)

![Example](images/example3.png)

### Setup

Create a Slack bot user from [here](https://my.slack.com/services/new/bot). You will receive an API token you can use. Set the `SLACK_API_TOKEN` environment variable and you should be good to go.

You can run this bot as below:

```shell
export SLACK_API_TOKEN="<SLACK_API_TOKEN>"
mix run --no-halt
```

A sample runner configuration looks like below:

```shell
export SLACK_API_TOKEN="<SLACK_API_TOKEN<"
export GITHUB_TOKEN="<GITHUB_TOKEN>"
export GOOGLE_API_KEY="<GOOGLE_API_KEY>"
export STANDUP_CHANNEL="scrum"
export GH_CHANNEL="github"
export PWN_CHANNEL="critical"
export UPTIME_CHANNEL="critical"

mix run --no-halt
```

### Scheduled Notifications

#### Github Pull Requests Watcher

You can configure github token by setting `GITHUB_TOKEN`. Also, you can pass list of repos to watch by updating [config](config/config.exs#L11-L17). There are bunch of other stuffs you can configure such as schedule (which follows cron format), slack channel and thresholds.

#### Standup Reminder

The standup reminder reminds us when its standup time. Our nature is that we either forget track of time or are too lazy to remember about it. This does not make you better at attending standup but is more of a satire for us. You can configure message and other bunch of stuffs on [config](config/config.exs#L5-L9)

#### Have I Been Pwned Checker

The hibp checker can check list of accounts against [haveibeenpwned.com](https://haveibeenpwned.com)
by using [ExPwned](https://github.com/techgaun/ex_pwned). You can set the appropriate configuration on config.exs.

```elixir
# config below runs hibp check once a day and posts if any found breach to configured slack channel
config :ex_mustang, ExMustang.Responders.Pwned,
  schedule: "0 * */1 * *",
  enabled: true,
  accounts: [
    "abc@example.com",
    "def@example.com"
  ],
  slack_channel: System.get_env("PWN_CHANNEL") || "general"
```

#### Uptime Monitoring

ExMustang supports simple uptime monitoring and can perform periodic uptime checks against given set of endpoints.
The config block looks like below:

```elixir
config :ex_mustang, ExMustang.Responders.Uptime,
  schedule: "*/5 * * *",
  enabled: true,
  endpoints: [
    [
      uri: "https://api.brighterlink.io/status", status_code: 200, content: ~s("msg":"ok"), method: "GET",
      content_type: "application/json", req_headers: [{"User-Agent", "ExMustang"}], timeout: 20_000
    ]
  ],
  slack_channel: System.get_env("UPTIME_CHANNEL") || "general"
```

The `endpoints` section is where you define the endpoints you're willing to perform uptime check against.

- `uri` - the endpoint to hit
- `status_code` - expected status code
- `content` - expected response body content (supports both string and regular expression)
- `method` - one of `GET`, `POST`, `PUT`, `PATCH` and `DELETE`
- `body` - body to send as part of `POST`, `PUT` or `PATCH` requests
- `content_type` - expected content type header value
- `req_headers` - list of tuples of request headers to send
- `timeout` - maximum timeout before request fails

#### Random Quote of the Day

ExMustang can send a quote as specified in schedule. You can configure schedule from config.exs
and you can specify `QUOTE_CHANNEL` env-var. If `QUOTE_CHANNEL` is not specified, the quotes are
sent to `general` channel by default.

### Responders

Currently, there are following responders and these should ideally work with any Hedwig adapters:

- `ExMustang.Responders.GMap` - gives you google maps search result for your search queries.
- `ExMustang.Responders.Pwned` -
- `ExMustang.Responders.Quote` - gives you random funny quote
- `ExMustang.Responders.Slap` - slaps another user
- `ExMustang.Responders.Time` - gives time on given timezone (without timezone, uses local timezone of machine ExMustang is running on)
- `ExMustang.Responders.TimeConvert` - converts given unix timestamp to ISO8601 format
- `ExMustang.Responders.EncodeDecode` - encoding/decoding tools
- `ExMustang.Responders.Isup` - checks if a site is up or not via isitup.org API
- `ExMustang.Responders.RandomInsult` - insults a user with random insults from randominsults.net
- `ExMustang.Responders.HTTPCat` - gets a cat for corresponding http status code
- `ExMustang.Responders.Howdoi` - tries to find answer for programming questions
- `ExMustang.Responders.CommitMsg` - Get random commit messages from http://whatthecommit.com/index.txt
- `ExMustang.Responders.CLIFu` - Gets clifu gems from http://www.commandlinefu.com
- `ExMustang.Responders.Whois` - performs a whois query using https://dnsquery.org
- `ExMustang.Responders.GitTip` - get a random git tip (with support for keywords)
- `ExMustang.Responders.Birthday` - send a happy birthday to the mentioned user

For Google Maps search, you have to set `GOOGLE_API_KEY` which has access to call google places api.

```shell
mustang help - Displays all of the help commands that mustang knows about.
mustang help <query> - Displays all help commands that match <query>.
gmap <search_term> - Replies with the information from google places/maps.
pwned <search_account> - Checks to see if an account has been breached or not
quote - Replies with a random quote.
slap - Slaps the user. Format: slap <username> | me
time - get time in a given timezone
unix2iso <unix_timestamp> - Converts given unix timestamp to ISO format (Auto-replies for values that look like timestamps)
b64encode <content> - base64 encoding of given text content
b64decode <content> - base64 decoding of given text content
isitup <domain> - checks if given domain is up or not
insult me|<username> - insults given username with random insults
httpcat <status_code> - get http status code cat for given value
howdoi <query> - tries to find solution on the given query
commitmsg - get a random commit message
clifu [search_word] - get clifu gem (gives random clifu if no keyword is passed)
whois <domain> - gives whois query for given domain
gittip [keyword] - Get a random git tip for given keyword
happy birthday <me|@user> - Send happy birthday message to the user mentioned
hdeploy <app-name> - Deploys configured branch to the given app
```

#### Heroku Deployment

For heroku deployment to work, you need to specify the following configurations:

- `HEROKU_GITHUB_TOKEN` or `GITHUB_TOKEN`
- `HEROKU_TOKEN` which is an API key you can get from [here](https://dashboard.heroku.com/account)
- mapping of apps in your config which is a tuple of heroku app name and map of repo and branch

An example config looks like below:

```elixir
config :ex_mustang, ExMustang.Responders.HerokuDeploy,
  github_token: System.get_env("HEROKU_GITHUB_TOKEN") || System.get_env("GITHUB_TOKEN"),
  token: System.get_env("HEROKU_TOKEN"),
  apps: [
    {"casa-core-stage", %{repo: "casaiq/core", branch: "master"}}
  ]
```

### Changelog

The detail changelog is available [HERE](CHANGELOG.md)

### About Mustang

Mustang (pronounced like `moostang` not like the `Ford Mustang`; I don't know phonetics) is my best friend whom I had to leave back in Nepal. After surviving [April 2015 Earthquake](https://en.wikipedia.org/wiki/April_2015_Nepal_earthquake), we lived together for a while and I had to leave him back in Nepal. Below is the pic post-earthquake.

![Mustang and Me](images/me_mustang.jpg)


CHANGELOG.md
# Change Log

## [0.3.0](https://github.com/techgaun/ex_mustang/tree/0.3.0) (2017-03-02)

[Full Changelog](https://github.com/techgaun/ex_mustang/compare/0.2.0...0.3.0)

**Closed issues:**

- fix timestamp extra intelligence [\#47](https://github.com/techgaun/ex_mustang/issues/47)
- git-tips [\#44](https://github.com/techgaun/ex_mustang/issues/44)
- add domain whois lookup feature [\#41](https://github.com/techgaun/ex_mustang/issues/41)
- small ideas to implement [\#26](https://github.com/techgaun/ex_mustang/issues/26)
- encode/decode [\#25](https://github.com/techgaun/ex_mustang/issues/25)
- Wish happy birthday if its someone's birthday [\#6](https://github.com/techgaun/ex_mustang/issues/6)
- Analyze slack online status of team members and predict if they are going to make it to standup [\#4](https://github.com/techgaun/ex_mustang/issues/4)

**Merged pull requests:**

- add heroku github deployment [\#53](https://github.com/techgaun/ex_mustang/pull/53) ([techgaun](https://github.com/techgaun))
- add birthday responder [\#52](https://github.com/techgaun/ex_mustang/pull/52) ([techgaun](https://github.com/techgaun))
- 1.4 warning fixes [\#48](https://github.com/techgaun/ex_mustang/pull/48) ([techgaun](https://github.com/techgaun))
- add git tip responder [\#46](https://github.com/techgaun/ex_mustang/pull/46) ([techgaun](https://github.com/techgaun))
- handle when record data is empty [\#43](https://github.com/techgaun/ex_mustang/pull/43) ([techgaun](https://github.com/techgaun))
- add whois responder [\#42](https://github.com/techgaun/ex_mustang/pull/42) ([techgaun](https://github.com/techgaun))

## [0.2.0](https://github.com/techgaun/ex_mustang/tree/0.2.0) (2016-12-17)
[Full Changelog](https://github.com/techgaun/ex_mustang/compare/0.1.0...0.2.0)

**Implemented enhancements:**

- add ability to specify name instead of identifier for slack channels [\#20](https://github.com/techgaun/ex_mustang/issues/20)

**Closed issues:**

- timestamp conversion [\#27](https://github.com/techgaun/ex_mustang/issues/27)
- simple uptime monitoring [\#18](https://github.com/techgaun/ex_mustang/issues/18)
- haveibeenpwned integration [\#16](https://github.com/techgaun/ex_mustang/issues/16)
- slapping ability to slap people [\#8](https://github.com/techgaun/ex_mustang/issues/8)

**Merged pull requests:**

- add clifu responder [\#40](https://github.com/techgaun/ex_mustang/pull/40) ([techgaun](https://github.com/techgaun))
- add commitmsg responder [\#39](https://github.com/techgaun/ex_mustang/pull/39) ([techgaun](https://github.com/techgaun))
- add howdoi responder [\#37](https://github.com/techgaun/ex_mustang/pull/37) ([techgaun](https://github.com/techgaun))
- add httpcat responder [\#36](https://github.com/techgaun/ex_mustang/pull/36) ([techgaun](https://github.com/techgaun))
- add random insults [\#35](https://github.com/techgaun/ex_mustang/pull/35) ([techgaun](https://github.com/techgaun))
- add isup responder and remove unneeded import [\#34](https://github.com/techgaun/ex_mustang/pull/34) ([techgaun](https://github.com/techgaun))
- add encode decode base64 [\#30](https://github.com/techgaun/ex_mustang/pull/30) ([techgaun](https://github.com/techgaun))
- auto reply for timestamps [\#29](https://github.com/techgaun/ex_mustang/pull/29) ([techgaun](https://github.com/techgaun))
- add unix ts to iso 8601 conversion [\#28](https://github.com/techgaun/ex_mustang/pull/28) ([techgaun](https://github.com/techgaun))
- add uptime monitoring [\#23](https://github.com/techgaun/ex_mustang/pull/23) ([techgaun](https://github.com/techgaun))
- add ability to disable schedules and add pwned schedule [\#22](https://github.com/techgaun/ex_mustang/pull/22) ([techgaun](https://github.com/techgaun))
- add ability to specify channel names instead of channel IDs [\#21](https://github.com/techgaun/ex_mustang/pull/21) ([techgaun](https://github.com/techgaun))
- add pwned responder [\#19](https://github.com/techgaun/ex_mustang/pull/19) ([techgaun](https://github.com/techgaun))
- Time responder [\#17](https://github.com/techgaun/ex_mustang/pull/17) ([techgaun](https://github.com/techgaun))
- Slap responder [\#15](https://github.com/techgaun/ex_mustang/pull/15) ([techgaun](https://github.com/techgaun))
- google maps search [\#14](https://github.com/techgaun/ex_mustang/pull/14) ([techgaun](https://github.com/techgaun))
- fix warnings with timex and hedwig [\#13](https://github.com/techgaun/ex_mustang/pull/13) ([techgaun](https://github.com/techgaun))

## [0.1.0](https://github.com/techgaun/ex_mustang/tree/0.1.0) (2016-10-27)
**Closed issues:**

- send fun quotes when user asks via "quote" [\#3](https://github.com/techgaun/ex_mustang/issues/3)
- Remind long waiting pull requests on watched list of github repos [\#2](https://github.com/techgaun/ex_mustang/issues/2)
- Remind everyone when its standup time [\#1](https://github.com/techgaun/ex_mustang/issues/1)



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

config :ex_mustang, ExMustang.Responders.Standup,
  time_of_day: "30 10 * * 1-5",
  slack_channel: System.get_env("STANDUP_CHANNEL") || "general",
  suffix: ["folks", "hackers", "peeps", "avengers"],
  msg: "Standup time",
  enabled: true

config :ex_mustang, ExMustang.Responders.Github,
  repos: ["techgaun/ex_mustang"],
  access_token: System.get_env("GITHUB_TOKEN"),
  schedule: "0 */1 * * *",
  slack_channel: System.get_env("GH_CHANNEL") || "general",
  created_time_threshold: 10800, # no old than 3 hours
  updated_time_threshold: 3600, # no old than 1 hour
  enabled: true

config :ex_mustang, ExMustang.Responders.Quote,
  quote_src: "files/quotes.txt",
  schedule: "*/1 * * * 1-5",
  slack_channel: System.get_env("QUOTE_CHANNEL") || "general",
  enabled: true

config :ex_mustang, ExMustang.Responders.Pwned,
  schedule: "59 23 */1 * *",
  enabled: true,
  accounts: [
    "abc@example.com",
    "def@example.com"
  ],
  slack_channel: System.get_env("PWN_CHANNEL") || "general"

config :ex_mustang, ExMustang.Responders.Uptime,
  schedule: "*/5 * * * *",
  enabled: true,
  endpoints: [
    [
      uri: "https://api.brighterlink.io/status", status_code: 200, content: ~s("msg":"ok"), method: "GET",
      content_type: "application/json", req_headers: [{"User-Agent", "ExMustang"}], timeout: 20_000
    ]
  ],
  slack_channel: System.get_env("UPTIME_CHANNEL") || "general"

config :ex_mustang, ExMustang.Responders.HerokuDeploy,
  github_token: System.get_env("HEROKU_GITHUB_TOKEN") || System.get_env("GITHUB_TOKEN"),
  token: System.get_env("HEROKU_TOKEN"),
  apps: [
    {"casa-core-stage", %{repo: "casaiq/core", branch: "master"}}
  ]

config :ex_mustang, ExMustang.Robot,
  adapter: Hedwig.Adapters.Slack,
  name: "mustang",
  aka: "/",
  token: System.get_env("SLACK_API_TOKEN"),
  responders: [
    {Hedwig.Responders.Help, []},
    {ExMustang.Responders.GMap, []},
    {ExMustang.Responders.Pwned, []},
    {ExMustang.Responders.Quote, []},
    {ExMustang.Responders.Slap, []},
    {ExMustang.Responders.Time, []},
    {ExMustang.Responders.TimeConvert, []},
    {ExMustang.Responders.EncodeDecode, []},
    {ExMustang.Responders.Isup, []},
    {ExMustang.Responders.RandomInsult, []},
    {ExMustang.Responders.HTTPCat, []},
    {ExMustang.Responders.Howdoi, []},
    {ExMustang.Responders.CommitMsg, []},
    {ExMustang.Responders.CLIFu, []},
    {ExMustang.Responders.Whois, []},
    {ExMustang.Responders.GitTip, []},
    {ExMustang.Responders.Birthday, []},
    {ExMustang.Responders.HerokuDeploy, []}
  ]

config :quantum, timezone: System.get_env("SYSTEM_TIME") || "America/Chicago"

config :ex_google,
  api_key: System.get_env("GOOGLE_API_KEY"),
  output: "json"


# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :ex_mustang, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:ex_mustang, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"
